### DISCUSSION TITLE
Defining a global variable in Cypress that can be used in all test files without creating a new one everytime

### DISCUSSION THREAD

**Discussion by mariamaslam** [cypress-io/cypress#28098] [Questions and Help]
I am very desperately looking for a solution to define a global variable and access it everything in all my test files. This is my commands.js. Here i am generating a random mobile number using faker.js. I am doing APIs automated testing but the APIs are executed in a sequence which means there should be one mobile number and one user ID that should be used throughout the test execution.

```
import { faker } from '@faker-js/faker';
Cypress.Commands.add('generateAndSetMobileNumber', () => {
  // Check if the mobile number is already set in the Cypress environment
  if (!Cypress.env('mobileNumber')) {
    const mobileNumber = faker.datatype.number({ min: 5700000000, max: 5899999999 });
    Cypress.env('mobileNumber', mobileNumber.toString());
  }
});

Cypress.Commands.add('createUser', () => {
  cy.generateRandomMobileNumber().then((mobileNumber) => {

    return cy.api({
      failOnStatusCode: false,
      method: 'POST',
      url: '/products/users/',
      body: {
        "countryCode": "971",
        "mobileNo": mobileNumber,
        "viewId": "createView",
        "agentCode": "",
        "language": "EN",
      }
    }).then((response) => {
      expect(response.status).to.eq(201);
      Cypress.env('prospectId', response.body.userId);
    });
  });
});
```
Now these commands are defined here and are working perfectly fine but when I use it in my test classes everything generates a new mobile number and new userID whereas I want one UserID and mobileNumber that can be used throughout the execution.

This is my test file

```
describe("Generating OTP", () => {
before(() => {
 cy.createUser() => {
     
    });
  it("Verify user is able to generate OTP by entering their mobile number", () => {
               UserId = Cypress.env('UserId');
               mobileNumber = Cypress.env('mobileNumber');


    cy.api({
      failOnStatusCode: false,
      method: 'POST',
      url: 'products/otp',
      body: {
        "action": "generate",
        "prospectId": prospectId,
        "countryCode": "971",
        "mobileNo": mobileNumber,
      }
    }).then((response) => {
      // Assert that the OTP was generated successfully
      expect(response.status).to.eq(200);
      expect(prospectId).to.have.length(5);
      // Assert that the OTP is valid
      expect(response.body.verified).to.eq(true);
    });
  });
});
```

I have tried multiple options as follow, but nothing seems to be working for me:

1. cypress-localstorage-commands plugin
2. setLocalStorage
3. getAllSessionStorage

I look forward to prompt support

**Comment by portyu9** [comment]
Hi @mariamaslam 

`Cypress.env()` is the wrong place for run-time state that needs to survive across spec files.

Each spec gets its own browser context, so browser-side state does not provide a reliable run-global variable. Also, in current Cypress (`16.x`), `Cypress.env()` has been removed.

For arbitrary values that need to persist between specs, keep them in Cypress's **Node process** and access them with `cy.task()`.

For example:

```js
// cypress.config.js
const { defineConfig } = require('cypress')

let runState = {}

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      on('task', {
        setRunState(values) {
          runState = { ...runState, ...values }
          return null
        },

        getRunState() {
          return runState
        },
      })
    },
  },
})
```

After creating the user:

```js
cy.task('setRunState', {
  mobileNumber,
  prospectId: response.body.userId,
})
```

and in another spec:

```js
cy.task('getRunState').then(({ mobileNumber, prospectId }) => {
  // use the values
})
```

Cypress specifically supports `cy.task()` for storing state between spec files.

There is one important limitation: that state belongs to **one Cypress process/machine**. If the specs run in parallel on multiple CI machines, they will not share it.

Since your API tests have a strict sequence, an even better design is usually either:

1. keep the dependent API workflow in one spec, or
2. create the required user independently for each spec.

That avoids making spec execution order part of the test contract.

Please mark this answer as accepted if it helped, thank you.
